diff --git a/languagepack/build-windows.sh b/languagepack/build-windows.sh
index 0e51c28..88a2145 100644
--- a/languagepack/build-windows.sh
+++ b/languagepack/build-windows.sh
@@ -168,7 +168,7 @@ EOT
     ssh $PG_SSH_WIN "cd $PG_PATH_WIN\\\\languagepack.$ARCH; mkdir -p $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20; cmd /c Perl_Build.bat $PG_PATH_WIN\\\\languagepack.$ARCH\\\\perl-5.20.3 $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20 PERL"
     ssh $PG_SSH_WIN "cd $PG_PATH_WIN\\\\languagepack.$ARCH; mkdir -p $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20; cmd /c Perl_Build.bat $PG_PATH_WIN\\\\languagepack.$ARCH\\\\perl-5.20.3 $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20 DBI"
     ssh $PG_SSH_WIN "cd $PG_PATH_WIN\\\\languagepack.$ARCH; mkdir -p $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20; cmd /c Perl_Build.bat $PG_PATH_WIN\\\\languagepack.$ARCH\\\\perl-5.20.3 $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20 DBD"
-
+    ssh $PG_SSH_WIN "cd $PG_PATH_WIN\\\\languagepack.$ARCH; mkdir -p $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20; cmd /c Perl_Build.bat $PG_PATH_WIN\\\\languagepack.$ARCH\\\\perl-5.20.3 $PG_LANGUAGEPACK_INSTALL_DIR_WIN\\\\Perl-5.20 IPC"
     # Generating/Updating liblzma.def file for Python Build
     if [ "$ARCH" = "windows-x32" ];
     then
diff --git a/languagepack/scripts/windows-x32/Perl_Build.bat b/languagepack/scripts/windows-x32/Perl_Build.bat
index cf5807d..0ac0d57 100644
--- a/languagepack/scripts/windows-x32/Perl_Build.bat
+++ b/languagepack/scripts/windows-x32/Perl_Build.bat
@@ -19,6 +19,7 @@ SET LINK=/SUBSYSTEM:CONSOLE,5.01
 IF "%vPerlModule%"=="PERL" GOTO PERL
 IF "%vPerlModule%"=="DBI" GOTO DBI
 IF "%vPerlModule%"=="DBD" GOTO DBD
+IF "%vPerlModule%"=="IPC" GOTO IPC
 GOTO END
 
 :PERL
@@ -43,5 +44,14 @@ CD %vPerlInstallDir%\bin
 ECHO ....Starting to Install DBD::PG....
 cpan install DBD::Pg
 ECHO ....End Install DBD::PG....
+GOTO END
+
+:IPC
+SET PATH=%vPerlInstallDir%\bin;%PATH%
+CD %vPerlInstallDir%\bin
+ECHO ....Starting to Install IPC::Run....
+cpan install IPC::Run
+ECHO ....End Install IPC::Run....
 
 :END
+
